### PR TITLE
Optimize per_group_transpose

### PR DESCRIPTION
**Pull request by laixinn**
## Motivation
Fusing shuffle_rows and per_group_transpose

## Benchmarking and Profiling
e2e performance on H20 by test_cutlass_moe.py and observe enhancement when batch size>=4096

```
Running benchmarks with TP size: 8
Testing batch sizes: [4096, 8192, 16384, 32768]
Model Config: {'num_experts': 256, 'topk': 8, 'hidden_size': 7168, 'shard_intermediate_size': 512, 'dtype': torch.bfloat16, 'block_shape': [128, 128]}

--- Batch Size: 4096 ---
Config: E=256, topk=8, H=7168, I_shard=512, dtype=torch.bfloat16, block_shape=[128, 128]
Warming up...
Benchmarking Cutlass fused_experts...
Benchmarking Cutlass shuffle fused_experts...
Benchmarking Triton fused_experts...
Cutlass fused_experts time: 2.590 ms (median) [2.589 - 2.591]
Shuffle fused_experts time: 2.576 ms (median) [2.575 - 2.577]
Triton  fused_experts time: 2.600 ms (median) [2.600 - 2.600]

--- Batch Size: 8192 ---
Config: E=256, topk=8, H=7168, I_shard=512, dtype=torch.bfloat16, block_shape=[128, 128]
Warming up...
Benchmarking Cutlass fused_experts...
Benchmarking Cutlass shuffle fused_experts...
Benchmarking Triton fused_experts...
Cutlass fused_experts time: 4.378 ms (median) [4.378 - 4.379]
Shuffle fused_experts time: 4.345 ms (median) [4.344 - 4.345]
Triton  fused_experts time: 4.697 ms (median) [4.696 - 4.697]

--- Batch Size: 16384 ---
Config: E=256, topk=8, H=7168, I_shard=512, dtype=torch.bfloat16, block_shape=[128, 128]
Warming up...
Benchmarking Cutlass fused_experts...
Benchmarking Cutlass shuffle fused_experts...
Benchmarking Triton fused_experts...
Cutlass fused_experts time: 7.906 ms (median) [7.904 - 7.907]
Shuffle fused_experts time: 7.830 ms (median) [7.829 - 7.830]
Triton  fused_experts time: 8.802 ms (median) [8.797 - 8.804]

--- Batch Size: 32768 ---
Config: E=256, topk=8, H=7168, I_shard=512, dtype=torch.bfloat16, block_shape=[128, 128]
Warming up...
Benchmarking Cutlass fused_experts...
Benchmarking Cutlass shuffle fused_experts...
Benchmarking Triton fused_experts...
Cutlass fused_experts time: 14.953 ms (median) [14.947 - 14.963]
Shuffle fused_experts time: 14.786 ms (median) [14.780 - 14.790]
Triton  fused_experts time: 16.952 ms (median) [16.949 - 16.953]
```

e2e accuracy test
```shell
# gsm8k: python3 bench_sglang.py --num-questions 1400 --parallel 200
Accuracy: 0.939
Invalid: 0.000
Latency: 135.454 s
Output throughput: 988.672 token/s
```

```shell
# mmlu: python3 bench_sglang.py --parallel 200
subject: abstract_algebra, #q:100, acc: 0.770
subject: anatomy, #q:135, acc: 0.889
subject: astronomy, #q:152, acc: 0.947
subject: business_ethics, #q:100, acc: 0.860
subject: clinical_knowledge, #q:265, acc: 0.917
subject: college_biology, #q:144, acc: 0.965
subject: college_chemistry, #q:100, acc: 0.660
subject: college_computer_science, #q:100, acc: 0.860
subject: college_mathematics, #q:100, acc: 0.790
subject: college_medicine, #q:173, acc: 0.873
subject: college_physics, #q:102, acc: 0.853
subject: computer_security, #q:100, acc: 0.890
subject: conceptual_physics, #q:235, acc: 0.949
subject: econometrics, #q:114, acc: 0.807
subject: electrical_engineering, #q:145, acc: 0.883
subject: elementary_mathematics, #q:378, acc: 0.944
subject: formal_logic, #q:126, acc: 0.817
subject: global_facts, #q:100, acc: 0.680
subject: high_school_biology, #q:310, acc: 0.958
subject: high_school_chemistry, #q:203, acc: 0.901
subject: high_school_computer_science, #q:100, acc: 0.950
subject: high_school_european_history, #q:165, acc: 0.891
subject: high_school_geography, #q:198, acc: 0.960
subject: high_school_government_and_politics, #q:193, acc: 0.984
subject: high_school_macroeconomics, #q:390, acc: 0.908
subject: high_school_mathematics, #q:270, acc: 0.778
subject: high_school_microeconomics, #q:238, acc: 0.962
subject: high_school_physics, #q:151, acc: 0.848
subject: high_school_psychology, #q:545, acc: 0.963
subject: high_school_statistics, #q:216, acc: 0.856
subject: high_school_us_history, #q:204, acc: 0.961
subject: high_school_world_history, #q:237, acc: 0.949
subject: human_aging, #q:223, acc: 0.870
subject: human_sexuality, #q:131, acc: 0.931
subject: international_law, #q:121, acc: 0.950
subject: jurisprudence, #q:108, acc: 0.898
subject: logical_fallacies, #q:163, acc: 0.926
subject: machine_learning, #q:112, acc: 0.795
subject: management, #q:103, acc: 0.932
subject: marketing, #q:234, acc: 0.949
subject: medical_genetics, #q:100, acc: 0.940
subject: miscellaneous, #q:783, acc: 0.953
subject: moral_disputes, #q:346, acc: 0.864
subject: moral_scenarios, #q:895, acc: 0.790
subject: nutrition, #q:306, acc: 0.922
subject: philosophy, #q:311, acc: 0.904
subject: prehistory, #q:324, acc: 0.944
subject: professional_accounting, #q:282, acc: 0.855
subject: professional_law, #q:1534, acc: 0.711
subject: professional_medicine, #q:272, acc: 0.949
subject: professional_psychology, #q:612, acc: 0.912
subject: public_relations, #q:110, acc: 0.809
subject: security_studies, #q:245, acc: 0.869
subject: sociology, #q:201, acc: 0.965
subject: us_foreign_policy, #q:100, acc: 0.950
subject: virology, #q:166, acc: 0.590
subject: world_religions, #q:171, acc: 0.930
Total latency: 164.891
Average accuracy: 0.875
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/layers/moe/cutlass_moe.py [hunk 1]
```diff
@@ -155,11 +156,25 @@ def cutlass_fused_experts_fp8(
...
+        else:
+            rep_a1_scales = shuffle_fp8_scale_hopper_moe_mn_major(
+                a1_scale,
+                expert_offsets[:-1],
+                problem_sizes1,
+                m * topk,
+                a_map,
+                expert_tokens_alignment=16,
+            )
+            w1_scale = w1_scale.contiguous()
```
Comment (HydraQYH) [inline python/sglang/srt/layers/moe/cutlass_moe.py]: There is a same line in line 165. Should move this line out of if statement.

File: python/sglang/srt/layers/moe/cutlass_moe.py [hunk 2]
```diff
@@ -155,11 +156,25 @@ def cutlass_fused_experts_fp8(
...
+            rep_a1_scales = per_group_transpose(rep_a1_scales, expert_offsets)
+            w1_scale = w1_scale.contiguous()
+        else:
+            rep_a1_scales = shuffle_fp8_scale_hopper_moe_mn_major(
+                a1_scale,
+                expert_offsets[:-1],
+                problem_sizes1,
+                m * topk,
+                a_map,
+                expert_tokens_alignment=16,
```
Comment (HydraQYH) [inline python/sglang/srt/layers/moe/cutlass_moe.py]: Should not hard code this. Are there other heuristic methods to automatically determine the factors of M?

File: python/sglang/srt/layers/quantization/fp8_kernel.py [hunk 3]
```diff
@@ -1633,3 +1633,142 @@ def triton_scaled_mm(
...
+    _shuffle_map_fp8_scale_hopper_moe_mn_major = shuffle_autotune(
+        _shuffle_map_fp8_scale_hopper_moe_mn_major
+    )
+
+def shuffle_fp8_scale_hopper_moe_mn_major(
+    a_s: torch.Tensor,
+    expert_offsets: torch.Tensor,
+    problem_sizes: torch.Tensor,
+    output_m: int,
+    shuffle_map: Optional[torch.Tensor] = None,
```
Comment (HydraQYH) [inline python/sglang/srt/layers/quantization/fp8_kernel.py:1916]: Can you explain the Shape & Stride & Meaning of shuffle_map?

**Review by HydraQYH — COMMENTED** [review]
Can you reply my comments?

File: python/sglang/srt/layers/moe/cutlass_moe.py (hunk 2 above)
Comment (laixinn) [reply python/sglang/srt/layers/moe/cutlass_moe.py]: alignment by element size should be a good idea that ignores tensor shape

**Comment by laixinn** [comment]
> Can you reply my comments?

Sorry for the late response. I have fixed most issues. Please check the alignment method.

File: python/sglang/srt/layers/quantization/fp8_kernel.py [hunk 4]
```diff
@@ -1633,3 +1633,143 @@ def triton_scaled_mm(
...
+    # index mapping from the shuffled destination (topk * M) to the source (M)
+    shuffle_map: Optional[torch.Tensor] = None,
+) -> torch.Tensor:
+    assert a_s.dim() == 2
+    assert a_s.is_contiguous(), "`A` is not contiguous"
+
+    k = a_s.shape[1]
+    sfa = torch.empty((output_m, k), device=a_s.device, dtype=a_s.dtype)
+    num_experts = problem_sizes.shape[0]
+    expert_tokens_alignment = a_s.element_size() * 8
```
Comment (HydraQYH) [inline python/sglang/srt/layers/quantization/fp8_kernel.py]: Why is it set up like this?

File: python/sglang/srt/layers/quantization/fp8_kernel.py [hunk 5]
```diff
@@ -1633,3 +1633,143 @@ def triton_scaled_mm(
...
+    E: tl.constexpr,
+    K: tl.constexpr,
+    M_ALIGNMENT: tl.constexpr,
+    BLOCK_M: tl.constexpr,  # tune
+    GROUP_M: tl.constexpr,
+):
+    pid = tl.program_id(0)
+
+    # swizzled order
+    width = GROUP_M * K
```
Comment (HydraQYH) [inline python/sglang/srt/layers/quantization/fp8_kernel.py:1836]: What are the benefits of this swizzle?
For example, E = 128, K = 56, GROUP_M = 4:
PID: 0, Expert : 0, K_Offset: 0
PID: 1, Expert : 1, K_Offset: 0
PID: 2, Expert : 2, K_Offset: 0
PID: 3, Expert : 3, K_Offset: 0
PID: 4, Expert : 0, K_Offset: 1
PID: 5, Expert : 1, K_Offset: 1
PID: 6, Expert : 2, K_Offset: 1
PID: 7, Expert : 3, K_Offset: 1
PID: 8, Expert : 0, K_Offset: 2
PID: 9, Expert : 1, K_Offset: 2
This will cause adjacent CTAs to store non-contiguous addresses.

**Review by HydraQYH — COMMENTED** [review]
Fusing shuffle & grouped transpose helps a lot. Please address the issues in the comments

File: python/sglang/srt/layers/quantization/fp8_kernel.py (hunk 4 above)
Comment (laixinn) [reply python/sglang/srt/layers/quantization/fp8_kernel.py]: Because alignment by element size must be safe regardless of tensor shape.

File: python/sglang/srt/layers/quantization/fp8_kernel.py (hunk 5 above)
Comment (laixinn) [reply python/sglang/srt/layers/quantization/fp8_kernel.py:1836]: Well, that is true. But the swizzle order turns out to bring better computation & memory throughputs, shown in ncu.

**Comment by laixinn** [comment]
Looks like i have solved all the issues. tl.multiple_of just sets a maximum value for divisibility, triton will finally determine a safe divisibility. cc @zhyncs @ispobock

**Comment by ispobock** [comment]
could you add e2e accuracy test results?

**Comment by ispobock** [comment]
lint ci should be fixed

**Comment by laixinn** [comment]
> lint ci should be fixed



> could you add e2e accuracy test results?

Done.

File: python/sglang/srt/layers/moe/cutlass_moe.py [hunk 6]
```diff
@@ -155,11 +156,24 @@ def cutlass_fused_experts_fp8(
...
+            rep_a1_scales = shuffle_rows(a1_scale, a_map, (m * topk, int(k / 128)))
+            rep_a1_scales = per_group_transpose(rep_a1_scales, expert_offsets)
+        else:
+            rep_a1_scales = shuffle_fp8_scale_hopper_moe_mn_major(
+                a1_scale,
+                expert_offsets[:-1],
+                problem_sizes1,
+                m * topk,
+                a_map,
+                M_ALIGNMENT=16,
```
Comment (HydraQYH) [inline python/sglang/srt/layers/moe/cutlass_moe.py]: Why hard code M_ALIGNMENT=16?

File: python/sglang/srt/layers/moe/cutlass_moe.py [hunk 7]
```diff
@@ -193,8 +207,17 @@ def cutlass_fused_experts_fp8(
...
-        w2_scale = w2_scale.contiguous()
+        if not use_shuffle:
+            a2_scale = per_group_transpose(a2_scale, expert_offsets)
+        else:
+            a2_scale = shuffle_fp8_scale_hopper_moe_mn_major(
+                a2_scale,
+                expert_offsets[:-1],
+                problem_sizes2,
+                m * topk,
+                M_ALIGNMENT=16,
```
Comment (HydraQYH) [inline python/sglang/srt/layers/moe/cutlass_moe.py]: Same as above comment.

File: python/sglang/test/test_cutlass_moe.py [hunk 8]
```diff
@@ -8,11 +8,12 @@
 
 from sglang.srt.layers.moe.cutlass_moe import cutlass_fused_experts_fp8
 from sglang.srt.layers.moe.fused_moe_triton.fused_moe import fused_experts
+from sglang.srt.layers.moe.moe_runner.base import MoeRunnerConfig
 
 
 def get_model_config(tp_size: int):
     config = AutoConfig.from_pretrained(
-        "deepseek-ai/deepseek-R1", trust_remote_code=True
+        "/workdir/huggingface.co/deepseek-ai/DeepSeek-V3-0324/", trust_remote_code=True
```
Comment (HydraQYH) [inline python/sglang/test/test_cutlass_moe.py]: Should not change this.

**Review by HydraQYH — COMMENTED** [review]
There are some other minor bugs that need to be fixed.

File: python/sglang/test/test_cutlass_moe.py [hunk 9]
```diff
@@ -105,6 +106,7 @@ def run_test(tp_size, batch_size, model_config, check=False):
         torch.rand(batch_size, topk, device="cuda", dtype=dtype), dim=-1
     )
     topk_ids = torch.randint(0, E, (batch_size, topk), dtype=torch.int32, device="cuda")
+    router_logits = torch.rand(batch_size, E, device="cuda", dtype=dtype)
```
Comment (HydraQYH) [inline python/sglang/test/test_cutlass_moe.py]: What is this used for? I don't see any use of this variable in the subsequent code.

**Review by HydraQYH — COMMENTED** [review]
Supplement.

File: python/sglang/srt/layers/quantization/fp8_kernel.py [hunk 10]
```diff
@@ -1700,3 +1700,144 @@ def triton_scaled_mm(
...
+    group_id = pid // width
+    group_size = tl.minimum(E - group_id * GROUP_M, GROUP_M)
+    expert_id = group_id * GROUP_M + (pid % group_size)
+    k_offset = (pid % width) // group_size
+
+    m = tl.load(problem_sizes + expert_id * 3)
+    m = tl.multiple_of(m, M_ALIGNMENT)
+    current_expert_offset = tl.load(expert_offsets + expert_id).to(tl.int64)
```
Comment (HydraQYH) [inline python/sglang/srt/layers/quantization/fp8_kernel.py:1843]: current_expert_offset can also be divided by M_ALIGNMENT?

**Review by HydraQYH — COMMENTED** [review]
Supplement secondly.

**Comment by laixinn** [comment]
> There are some other minor bugs that need to be fixed.

> Supplement.

> Supplement secondly.

Issues fixed. @HydraQYH @ispobock

**Comment by HydraQYH** [comment]
After the https://github.com/sgl-project/sglang/pull/9559 is merged, this PR will be invalid because we don't need per_group_transpose after CUTLASS 4.2. But we still appreciate your code contribution.
